src/VisualStudio/Core/Def/Implementation/ProjectSystem/DocumentProvider.cs

```diff
@@ -381,24 +563,39 @@ private void OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, IVsHierar
 
         protected virtual void OnHierarchyChanged(uint docCookie, IVsHierarchy pHierOld, uint itemidOld, IVsHierarchy pHierNew, uint itemidNew, bool itemidChanged)
         {
+            AssertIsForeground();
+
+            var documentBuilder = ArrayBuilder<StandardTextDocument>.GetInstance();
             List<DocumentKey> documentKeys;
-            if (_docCookiesToOpenDocumentKeys.TryGetValue(docCookie, out documentKeys))
+
+            lock (_gate)
             {
-                foreach (var documentKey in documentKeys)
+                if (_docCookiesToOpenDocumentKeys.TryGetValue(docCookie, out documentKeys))
                 {
-                    var document = _documentMap[documentKey];
-                    var currDocHier = document.Project.Hierarchy;
-
-                    if (currDocHier == pHierNew)
+                    foreach (var documentKey in documentKeys)
                     {
-                        documentKey.HostProject.Workspace.OnDocumentContextUpdated(document.Id, document.GetOpenTextContainer());
+                        documentBuilder.Add(_documentMap[documentKey]);
                     }
                 }
             }
+
+            for (var i = 0; i < documentBuilder.Count; i++)
+            {
+                var document = documentBuilder[i];
+                var currDocHier = document.Project.Hierarchy;
+
+                if (currDocHier == pHierNew)
+                {
+                    var documentKey = documentKeys[i];
```

Comment:
Consider doing all of this under the gate and then just moving the notification out. It might be easier to reason about. (It took be about a minute to figure out what's up with all the copying of state around.)